packages/jsii-rosetta/lib/languages/java.ts

```diff
@@ -0,0 +1,629 @@
+import ts = require('typescript');
+import { DefaultVisitor } from './default';
+import { AstRenderer } from '../renderer';
+import { NO_SYNTAX, OTree } from '../o-tree';
+import { builtInTypeName, mapElementType, typeWithoutUndefinedUnion } from '../typescript/types';
+import { isReadOnly, matchAst, nodeOfType, quoteStringLiteral, visibility } from '../typescript/ast-utils';
+import { ImportStatement } from '../typescript/imports';
+import { jsiiTargetParam } from '../jsii/packages';
+
+interface JavaContext {
+  /** @default false */
+  readonly ignorePropertyPrefix?: boolean;
+
+  /** @default true */
+  readonly convertPropertyToGetter?: boolean;
+
+  readonly insideTypeDeclaration?: InsideTypeDeclaration;
+
+  readonly inKeyValueList?: boolean;
+
+  /**
+   * Used when rendering a JavaScript object literal that is _not_ for a struct -
+   * we render that as a Map in Java.
+   *
+   * @default false
+   */
+  readonly identifierAsString?: boolean;
+
+  /**
+   * Used when rendering a JavaScript object literal that is for a struct -
+   * maps to a Builder in Java.
+   *
+   * @default false
+   */
+  readonly stringLiteralAsIdentifier?: boolean;
+}
+
+interface InsideTypeDeclaration {
+  readonly typeName: ts.Node | undefined;
+}
+
+type JavaRenderer = AstRenderer<JavaContext>;
+
+export class JavaVisitor extends DefaultVisitor<JavaContext> {
+  public readonly defaultContext = {};
+
+  public mergeContext(old: JavaContext, update: Partial<JavaContext>): JavaContext {
+    return Object.assign({}, old, update);
+  }
+
+  public importStatement(importStatement: ImportStatement): OTree {
+    const namespace = this.lookupModuleNamespace(importStatement.packageName);
+    if (importStatement.imports.import === 'full') {
+      return new OTree([`import ${namespace}.*;`], [], { canBreakLine: true });
+    }
+    return new OTree(
+      [],
+      importStatement.imports.elements.map(importEl => `import ${namespace}.${importEl.sourceName};`),
+      { canBreakLine: true, separator: '\n' },
+    );
+  }
+
+  /*
+   * Because structs are represented in Java as interfaces with a JsiiProxy implementation,
```

Comment:
But what if a user wants to write a struct themselves in Java, to be used in Java? Then there will be no `JsiiProxy`. They would probably write a class to hold their "props". See the C# generator.

This translator is about code the user would write, not mirroring what jsii does for them.

Reply:
Can you elaborate on what the translation should be? Let's say we have

```ts
interface MyStruct {
  readonly a: string;
  b: number;
}
```

in TypeScript. Should I generate

a) an interface with getters (and setters for non-readonly fields)?
b) a class with a constructor taking all fields as arguments, and just assigning them in that constructor?
c) a class that has a nested static Builder class that is used to create it, and the constructor of the class itself is private, similar to what JSII is doing?
d) a  simple class with just the fields, and Lombok annotations (`@Builder`, `@Data`, etc.)?
e) something else...?

Reply:
Yeah, good question. I'd like the syntactically most lightweight thing that roughly expresses the intent. I have a feeling the builder would add a lot of noise to make it work nicely at the consumer site, so I think I'd prefer either b or d (or b as a POJO).

I guess between those it depends on how reasonable it is for people to have Lombok installed. If not, I think POJOs are a reasonable compromise? Or honestly, publicly accessible fields without getters or setters (is what I did in C#). In fact, that is what structs are, just a bag of data. `readonly` doesn't really come into it, because this is only for structs and all fields are readonly by definition.

I can't really tell you the solution here, only what I value: readability and simplicity. I don't want all the syntax of a "super up-to-snuff" class definition get in the way of the rest of the example. People know how to translate from plain public attributes or POJOs to their preferred immutable/builder/deserialize/Lombok flavors that apply.

I think non-final public variables are the simplest solution. Only thing that gives me slight pause there is that people will get stuck up parroting how "that's an anti-pattern" to us and missing the forest for the trees.

Wrap those suckers up in getters and setters that don't do anything other than manipulating those same members (and crucially, never will, either!) and suddenly nobody bats an eye. Go figure.

I need your opinion here. Do what seems right to you.

Reply:
OK. I've settled on private fields + getters and fluent setters, so creating instances of the class looks something like this:

```java
new MyStruct()
        .a("a")
        .b(42);
```

, which seems the closest we can get in Java to JavaScript's object literal syntax without generating a separate Builder class, which would be very verbose, like you said.

Let me know if this works!